lib/fixed-points.nix

```diff
@@ -110,4 +110,94 @@ rec {
     fix' rattrs // {
       ${extenderName} = f: makeExtensibleWithCustomName extenderName (extends f rattrs);
    };
+
+  /*
+    Creates an overridable attrset with encapsulation.
+
+    This is like `makeExtensible`, but only the `public` attribute of the fixed
+    point is returned.
+
+    Synopsis:
+
+        r = encapsulate (final@{extend, ...}: {
```

Comment:
Putting `extend` in `final` is conceptually not valid. It should be a separate parameter.

(It's the kind of code you end up with when trying to stick too close to past patterns, as in #119942)

Reply:
> Putting `extend` in `final` is conceptually not valid. It should be a separate parameter.

I believe this comes from the discussion in https://github.com/NixOS/nixpkgs/pull/157056#discussion_r809325176? If so, I think the concerns there have been cleared

Reply:
Indeed your comment seems to be about the same kind of problem.

I think I was setting the bar too high when I started this thread. What's the alternative? Add an extra parameter?
Reserving a name in the attrset of protected attributes is not pretty, but gets the job done quite well, without burdening the user with extra syntax.

The extra parameter will cause it to look like this:

```nix
lib.encapsulate (extend: this: {
  # ...
  public.withSomething = s: extend (extend: this: super: {
    something = s;
  });
});
```

Doesn't seem like a usability improvement, especially if you're not actually going to use `extend`.

`this.extend` seems better now, even if it means that users can't use that name.
That's probably for the best though, because we should having too many overriding mechanisms. If they do need to use that name, the problem is that they have one too many overriding mechanisms.

Reply:
I suppose a more realistic alternative is

```nix
lib.encapsulate ({ this, extend, ... }: {
  # ...
  public.withSomething = s: extend ({ this, extend, super, ... }: { something = s; }); });
}
```

We don't have renaming for such parameters though, which can make nested use awkward, as seen with not-just-data submodules, which have the same problem.

<details><summary>rant and analogy with module system</summary>

Also before you know it, someone will want to extend that attrset, and we'll have `encapsulateWith { specialArgs = ...; }`, even though `super` can fill the role of `specialArgs`. They're actually oddly similar.
Overlays have `//` for "lateral" composition (idk, is that term taken?), whereas the module system has option merging for that. `extendModules` then lines up with `extends`. `extendModules` relies on the lexical scope (ie `config`) to provide the "super" context, whereas in overlays it's an explicit parameter. `specialArgs` is effectively an extension of all modules' lexical scope, so it's a close analog of `super`.
This seems like an overreaction though.

</details>

The lack of nested "formals" (`{ this@{ foo, ... }, ... }: <body>`) could be a feature though, because then it's not possible to do the equivalent of what's currently `lib.encapsulate ({ foo, ... }: <body>)`, which doesn't work, because such lambdas are strict (a phenomenon very relevant to [`lazyFunction`](https://github.com/NixOS/nixpkgs/pull/194514), for some context).